system/include/emscripten/promise.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2023 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#pragma once
+
+#include <stdlib.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// An opaque handle to a JS Promise object.
+typedef struct em_promise_t {
+  void* handle;
+} em_promise_t;
```

Comment:
How about just `typedef uint32_t em_promise_t`  ?   Or `typedef void* em_promise_t`  (i.e. Why does it need to be struct type?)

Reply:
I originally used `uint32_t` for the inner handle, but using the API often requires casting these handles to `void*`, so my test code was covered with ugly `*result = (void*)(uintptr_t)promise.handle;` lines. I changed the handle to be void* so that those lines can be much nicer: `*result = promise.handle`.

Reply:
Maybe put those casts into helpers/macros to if you think they ugly?  Are those casts needed when you want pass a promise as user data?   Aren't those cases expected whenever one wants to pass non-pointer-sized data as user data?   One could make the argument for make all types pointer sized based on that argument couldn't you?

I worry that we it will hard to change the underlying type here once we decide on its width.   I guess there is any way to avoid exposing its width.

Reply:
Also, why `*result = (void*)(uintptr_t)promise.handle` and hot `*result = (void*)(uintptr_t)promise`?  Isn't the em_promise type itself that thing you want to pass as the user data, and not the inner handle?

Reply:
Because I don't believe it's possible to cast a struct to a number or pointer, even if they're the same size (but I haven't checked).

Reply:
> Are those casts needed when you want pass a promise as user data?

Yes, but more importantly when you want to return from a promise callback with `EM_PROMISE_MATCH` or `EM_PROMISE_MATCH_RELEASE` and need to write the promise handle to the `result` pointer.

Reply:
Hmm.. makes me think maybe we use different types argument for different result types.  e.g and extra paramater called such as `em_handle_t* match_result_or_null`? 

Maybe its fine, but all that casting seems dirty.

Reply:
Macros or helpers could help, but I would prefer to avoid them as well. What would you think of doing `typedef struct _em_promise* em_promise_t` similar to how `pthread_t` works for C? For C at least, that would avoid the need for any casts and also avoid `.handle`, since all pointers are assignable to `void*`. C++ would still need explicit casts to `void*`, but that's probably ok (or we can do a different typedef in C++, like `pthread_t` has).